### PR TITLE
feat(config-loader): support a rule written by ESM

### PR DESCRIPTION
**Pull request by azu**
Currently, secretlint can not load a rule that is written by ESM.

This PR aims to allow to load secretlint rule as  ESM modules.

## Blocker

TypeScript does not support `import()` in CommonJS

- [Add flag to not transpile dynamic `import()` when module is `CommonJS` · Issue #43329 · microsoft/TypeScript](https://github.com/microsoft/TypeScript/issues/43329)

`__esModule` interop is broken.
We need to pick `.default.default`. It is ugly.

- [ ] we will drop to use default export in rule

### CODE REVIEW

File: packages/@secretlint/config-loader/src/index.ts [hunk 1]
```diff
@@ -8,9 +8,18 @@ import {
 } from "@secretlint/types";
 import { secretLintProfiler } from "@secretlint/profiler";
 import { SecretLintModuleResolver } from "./SecretLintModuleResolver";
-import { moduleInterop } from "@textlint/module-interop";
 import { validateConfig, validateRawConfig } from "@secretlint/config-validator";
 
+export function moduleInterop<T>(moduleExports: T): T {
+    if (moduleExports && (moduleExports as any).__esModule) {
+        return (moduleExports as any).default?.default!;
```
Comment (azu) [inline packages/@secretlint/config-loader/src/index.ts:15]: `module.exports.default = 1`

is `default.default`.
`__esModule` is broken.
image

**Comment by jimmywarting** [comment]
tip: just something that helped me with async imports:
[TypeScript/WebPack rewrites my async imports to require in CJS projects](https://github.com/node-fetch/node-fetch/issues/1279#issuecomment-915063354)

File: packages/@secretlint/config-loader/src/index.ts [hunk 2]
```diff
@@ -8,9 +8,18 @@ import {
...
+    }
+    return moduleExports && (moduleExports as any).default ? (moduleExports as any).default! : moduleExports;
+}
+
+// FIXME: https://github.com/microsoft/TypeScript/issues/43329
+// module: node12 will be replace it
+const _importDynamic = new Function("modulePath", "return import(modulePath)");
```
Comment (azu) [inline packages/@secretlint/config-loader/src/index.ts:22]: Dynamic Import hack in CommonJS + TypeScript.

- [ ] https://github.com/microsoft/TypeScript/issues/43329 will fix it

**Comment by azu** [comment]
Yes. We have used `_importDynamic` workaround.

File: packages/@secretlint/config-loader/src/index.ts (hunk 2 above)
Comment (azu) [reply packages/@secretlint/config-loader/src/index.ts:32]: This workaround should be removed in the future

- CSP Issue
- Performance Issue
- Type Issue

**Pull request merged**